pkg/castemplate/v1alpha1/kubernetes_test.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2019 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/pkg/errors"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func fakeGetClientset() (cs *clientset.Clientset, err error) {
+	return &clientset.Clientset{}, nil
+}
+
+func fakeGetfn(cs *clientset.Clientset, name string,
+	opts metav1.GetOptions) (*apis.CASTemplate, error) {
+	return &apis.CASTemplate{}, nil
+}
+
+func fakeGetErrfn(cs *clientset.Clientset, name string,
+	opts metav1.GetOptions) (*apis.CASTemplate, error) {
+	return &apis.CASTemplate{}, errors.New("some error")
+}
+
+func fakeSetClientset(k *Kubeclient) {
+	k.clientset = &clientset.Clientset{}
+}
+
+func fakeSetNilClientset(k *Kubeclient) {
+	k.clientset = nil
+}
+
+func fakeGetNilErrClientSet() (clientset *clientset.Clientset, err error) {
+	return nil, nil
+}
+
+func fakeGetErrClientSet() (clientset *clientset.Clientset, err error) {
+	return nil, errors.New("Some error")
+}
+
+func fakeClientSet(k *Kubeclient) {}
+
+func TestWithDefaults(t *testing.T) {
+	tests := map[string]struct {
+		getFn              getFunc
+		getClientsetFn     getClientsetFunc
+		expectGet          bool
+		expectGetClientset bool
+	}{
+		// The current implementation of WithDefaults method can be
+		// tested using these two combinations only.
+		"When mockclient is empty": {nil, nil, false, false},
+		"When mockclient contains all of them": {fakeGetfn,
+			fakeGetClientset, false, false},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			fc := &Kubeclient{}
+			fc.get = mock.getFn
+			fc.getClientset = mock.getClientsetFn
+
+			fc.withDefaults()
+			get := (fc.get == nil)
+			if get != mock.expectGet {
```

Comment:
This is giving me hard time to understand. Can we simplify?
```go
if mock.expectGet && fc.get == nil {
  // ...
}
```

Reply:
done